### PR TITLE
Support for transformEvent to return an array

### PR DESCRIPTION
**Pull request by psrikanth88**
## Description of the change

> Description here

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

**Review by lokey — APPROVED** [review]

File: util/ivmFactory.js [hunk 1]
```diff
@@ -38,31 +38,64 @@ async function createIvm(code, libraryVersionIds) {
...
-              const currMsgId = ev.messageId;
-              let transformedEvent = await transformEvent(ev, metadata);
-              return {transformedEvent, metadata: eventsMetadata[currMsgId] || {}};
+              let transformedOutput = await transformEvent(ev, metadata);
+              // if func returns null/undefined drop event
+              if (transformedOutput === null || transformedOutput === undefined) return;
+              if (Array.isArray(transformedOutput)) {
+                const producedEvents = [];
+                let encounteredError = false;
+                transformedOutput.every(e => {
```
Comment (sivashanmukh) [inline util/ivmFactory.js:74]: Why use `.every` when you don't collect it's return?

File: util/ivmFactory.js [hunk 2]
```diff
@@ -38,31 +38,64 @@ async function createIvm(code, libraryVersionIds) {
       const eventMessages = events.map(event => event.message);
       const eventsMetadata = {};
       events.forEach(ev => {
-        eventsMetadata[ev.message.messageId] = ev.metadata;
+        eventsMetadata[ev.metadata ? ev.metadata.messageId : ev.message.messageId] = ev.metadata;
```
Comment (sivashanmukh) [inline util/ivmFactory.js:41]: Why was this required?

File: util/ivmFactory.js [hunk 3]
```diff
@@ -38,31 +38,64 @@ async function createIvm(code, libraryVersionIds) {
...
       });
 
+      const isObject = (o) => Object.prototype.toString.call(o) === '[object Object]';
```
Comment (sivashanmukh) [inline util/ivmFactory.js:44]: Is this safe to use instead `o.constructor == Object`
Just because it looks better. Not important though

File: util/ivmFactory.js [hunk 4]
```diff
@@ -38,31 +38,64 @@ async function createIvm(code, libraryVersionIds) {
...
+              } 
+              outputEvents.push({transformedOutput, metadata: eventsMetadata[currMsgId] || {}});
+              return;
             } catch (error) {
               // Handling the errors in versionedRouter.js
-              return {error: error.toString(), metadata: metadata(ev)};
+              return outputEvents.push({error: error.toString(), metadata: eventsMetadata[currMsgId] || {}});
             }
           }));
-          outputEvents = outputEvents.filter(e => (e.transformedEvent != null || e.error));
```
Comment (sivashanmukh) [inline util/ivmFactory.js:65]: Filter out on null is still required right?

**Review by sivashanmukh — COMMENTED** [review]
Just to cross-verify: The changes in the test file are just the addition of new tests and nothing else right?

File: util/ivmFactory.js (hunk 1 above)
Comment (psrikanth88) [reply util/ivmFactory.js:74]: to break when non object is found

File: util/ivmFactory.js (hunk 2 above)
Comment (psrikanth88) [reply util/ivmFactory.js:41]: yeah, it was not actually required. was changing everywhere to read from metadata instead of message. will change back

File: util/ivmFactory.js (hunk 3 above)
Comment (psrikanth88) [reply util/ivmFactory.js:44]: `o.constructor` returns a function. comparing it would be difficult. this seems to be recommended by more in stackoverflow etc

File: util/ivmFactory.js (hunk 4 above)
Comment (psrikanth88) [reply util/ivmFactory.js:65]: we are not pushing null values above. returning error/empty outputEvents in such cases
